nemo/collections/tts/modules/submodules.py

```
@@ -410,3 +419,284 @@ def forward(self, forward_input: Tuple[torch.Tensor, torch.Tensor]):
+        return x.unsqueeze(1)
+
+
+class SpeakerEncoder(torch.nn.Module):
+    """
+    class SpeakerEncoder represents speakers representation. This module can combine GST (global style token)
+    based speaker embeddings, speaker embeddings from speaker verificaiton models and lookup table speaker
+    embeddings.
+    """
+
+    def __init__(self, gst_module=None, sv_projection_module=None, lookup_emb_projection_module=None):
+        """
+        Constructor.
+        Parameters:
+            - gst_module: Neural module to get GST based speaker embedding
+            - sv_projection_module: Neural module to get Speaker Verification based speaker embedding
+            - lookup_emb_projection_module: Neural module to project lookup speaker embedding
+        """
+        super(SpeakerEncoder, self).__init__()
+        self.gst_module = gst_module
+        self.sv_projection_module = sv_projection_module
+        self.lookup_emb_projection_module = lookup_emb_projection_module
+
+    def forward(self, spk_emb, reference_spec=None, reference_spec_lens=None, speaker_embedding=None):
+        x = spk_emb
+        # Project lookup speaker embedding
+        if self.lookup_emb_projection_module is not None:
+            x += self.lookup_emb_projection_module(spk_emb)
+
+        # Get GST based speaker embedding
+        if self.gst_module is not None and reference_spec is not None and reference_spec_lens is not None:
+            ref_spec_mask = (
+                torch.arange(reference_spec_lens.max())
+                .to(reference_spec.device)
+                .expand(reference_spec_lens.shape[0], reference_spec_lens.max())
+                < reference_spec_lens.unsqueeze(1)
+            ).unsqueeze(2)
+            x += self.gst_module(reference_spec, ref_spec_mask).unsqueeze(1)
+
+        # Project speaker embedding from Speaker Verification based
+        if self.sv_projection_module is not None and speaker_embedding is not None:
```

Should we throw an error if sv_projection_module is not None but speaker_embedding is None?

---

Add the check logic in models/fastpitch.py https://github.com/NVIDIA/NeMo/blob/tts_adapters_cln/nemo/collections/tts/models/fastpitch.py#L151-L176